erasurecoding/src/lib.rs

```diff
@@ -0,0 +1,122 @@
+                let end = start + shard_size as usize;
+                output_slice[start..end].copy_from_slice(shard);
+            }
+            0
+        }
+        Err(_) => -1,
+    }
+}
+
+/// Reed-Solomon decode function for FFI
+/// Reconstructs original data shards from combination of original and recovery
+/// shards. Note that only missing original shards are recovered. Any original
+/// shards supplied are not recovered since we already have them.
+/// Parameters:
+/// - original_shards_count: number of original shards
+/// - recovery_shards_count: number of recovery shards
+/// - shard_size: size of each shard in bytes
+/// - original_shards_data: original flattened shards
+/// - original_shards_len: length of original_shards_data
+/// - original_shards_indexes: indexes of original_shards_data
+/// - recovery_shards_data: recovery flattened shards
+/// - recovery_shards_len: length of recovery_shards_data
+/// - recovery_shards_indexes: indexes of recovery_shard_data
+/// - recovered_shards_out: buffer for recovered missing original shards
+/// - recovered_shards_indexes_out: buffer for indexes of recovered original shards
+/// Returns 0 on success, -1 on error
+#[no_mangle]
+pub unsafe extern "C" fn reed_solomon_decode(
+    original_shards_count: size_t,
+    recovery_shards_count: size_t,
+    shard_size: size_t,
+    original_shards: *const u8,
+    original_shards_len: size_t,
+    original_shards_indexes: *const size_t,
+    recovery_shards: *const u8,
+    recovery_shards_len: size_t,
+    recovery_shards_indexes: *const size_t,
+    recovered_shards_out: *mut u8,
+    recovered_shards_indexes_out: *mut size_t,
+) -> c_int {
+    // Create original shard pairs
+    let orig_data = slice::from_raw_parts(original_shards, original_shards_len);
+    let orig_indexes =
+        slice::from_raw_parts(original_shards_indexes, original_shards_len / shard_size);
+    let original_inputs = orig_indexes
+        .iter()
+        .zip(orig_data.chunks(shard_size))
+        .map(|(&idx, chunk)| (idx, chunk));
+
+    // Create recovery shard pairs
+    let rec_data = slice::from_raw_parts(recovery_shards, recovery_shards_len);
+    let rec_indexes =
+        slice::from_raw_parts(recovery_shards_indexes, recovery_shards_len / shard_size);
+    let recovery_inputs = rec_indexes
+        .iter()
+        .zip(rec_data.chunks(shard_size))
+        .map(|(&idx, chunk)| (idx, chunk));
+
+    match reed_solomon_simd::decode(
+        original_shards_count,
+        recovery_shards_count,
+        original_inputs,
+        recovery_inputs,
+    ) {
+        Ok(restored) => {
+            let shards_recovered_out =
+                slice::from_raw_parts_mut(recovered_shards_out, restored.len() * shard_size);
+
+            let shards_recovered_indexes_out =
+                slice::from_raw_parts_mut(recovered_shards_indexes_out, restored.len());
+
+            for (i, (&shard_index, shard_data)) in restored.iter().enumerate() {
```

Comment:
I don't know how long can be `restored`, but if there are many entries, it might be worth considering a parallel iterator. (This can also be applied when copying slices in the `encode` function).

Reply:
We'll be dealing around 342 * 2 bytes most of the time, so parallel iterators probably aren't worth it yet.